### PR TITLE
Add a MkDocs Material documentation website

### PR DESCRIPTION
**Pull request by BrunoB81HK**
I'm so glad to have found this cookbook. It is a great collection of snippets and I believe that a cookbook, with many more snippets could become an invaluable central tool for every ROS developers.

The only thing that, in my opinion, is missing is a better way to navigate the cookbook.

That being said, in this pull request, I added a MkDocs Material documentation website as well as a simple devcontainer to facilitate the development. To achieve it, I had to restructure the repo. The current design/layout is rough, I patched something quickly to have something fast.

I'm looking forward to your response and comments!

Thanks, 

Bruno

### CODE REVIEW

**Comment by mikeferguson** [comment]
I'm currently traveling, this will take me a little bit of time to look at

File: README.md [hunk 1]
```diff
@@ -1,62 +1,62 @@
...
-
-The contents of this cookbook are placed in the Public Domain through the CC0 license.
-You can use code snippets in your code without attribution and without impact to your
-license choice. This cookbook is provided as-is and without any warranties of any kind.
-See the full text of the
-[CC0 license](https://creativecommons.org/publicdomain/zero/1.0/legalcode).
-
-While not required by license, if you want to copy the entire cookbook somewhere, please
-give some attribution.
+# ROS2 Cookbook
```
Comment (mikeferguson) [inline README.md]: any idea why GitHub thinks this line has changed? Is your editor maybe not using unix line endings? If so, that's making the diff much larger than it needs to be

**Review by mikeferguson — COMMENTED** [review]
What's the deployment story here? Would this end up rendering into a GitHub.io webpage at that point?

File: README.md (hunk 1 above)
Comment (BrunoB81HK) [reply README.md]: Yeah that is exactly what happened, I solved it in d469cd6a93cf1d6d2c3607c5ae67e9e5b93aa5c4

**Comment by BrunoB81HK** [comment]
> What's the deployment story here? Would this end up rendering into a GitHub.io webpage at that point?

This is a good question. I must admit that I haven't thought about it that much. GitHub Page is probably the best and quickest solution. I'm not sure how to set it up, but after a quick look at the documentation, it doesn't seem to difficult.

Edit: I just tried to deploy it from my forked repo, it is a single shell command (`mkdocs gh-deploy`). You can check it out [here](https://brunob81hk.github.io/ros2_cookbook/).

**Comment by mikeferguson** [comment]
Ok - I've got a couple thoughts here:

 * I really don't like moving all the files into a directory - a lot of these pages are very highly ranked in google right now (some of them actually show up BEFORE docs.ros.org, especially where google can't seem to show anything newer than Foxy).
 * Given that - and the fact that things need to go into a gh-pages branch anyways in order to be published to web - my suggestion here is to basically add an extra step that temporarily copies everything into a docs folder before running mk-docs during the build procedure (that way, we can continue to have the structure we have today, and also support building mk-docs).

Does that make sense? If so, the steps we'll need to do:

 * Add the copy script and revert the move of all the files (you)
 * I think I need to adjust some settings on my GitHub account to make gh-pages work properly for my repos
 * Once we merge this, I can do a first deploy to gh-pages manually and then add a GitHub actions that does the copy + mk-docs

**Comment by BrunoB81HK** [comment]
> * I really don't like moving all the files into a directory - a lot of these pages are very highly ranked in google right now (some of them actually show up BEFORE docs.ros.org, especially where google can't seem to show anything newer than Foxy).

This is an excellent point. I tried to use the structure directory as is, but it was not possible with MkDocs as I understood it. I Like your idea, but I may have a simple idea I want to try before using a script. I would use a script only as a last resort mainly because I would like to stick to MkDocs' standards as much as possible.

I'll try to experiment on this today.

Thanks a lot for your time!

**Comment by BrunoB81HK** [comment]
I was finally able to make it work. I havent fully checked for typos and the formatting could be improved, but the original structure is now kept.

It essentially use a python script to create the needed pages at the right place, a bit like you suggested. 

Let me know what you think!

**Comment by BrunoB81HK** [comment]
@mikeferguson Any updates on this?

**Comment by chenyenru** [comment]
Hi I'm interested in helping. Please let me know if there are any updates.

**Comment by BrunoB81HK** [comment]
@mikeferguson Have you had the time to review the changes I made?

**Comment by mikeferguson** [comment]
Superseded by https://github.com/mikeferguson/ros2_cookbook/pull/14 - I fixed the merge conflicts and touched up some left over issues in the README - already deployed that branch to https://mikeferguson.github.io/ros2_cookbook/

**Comment by mikeferguson** [comment]
Updated branch is merged and deployed - I think there are still some rough spots around the edges, so feel free to open follow on PRs as you find issues.

**Comment by BrunoB81HK** [comment]
Great! I will check it out!

**Pull request closed without merging**